buffer/src/main/java/io/netty/buffer/api/Send.java

```diff
@@ -0,0 +1,130 @@
+     */
+    static <T extends Resource<T>> Send<T> sending(Class<T> concreteObjectType, Supplier<? extends T> supplier) {
+        return new Send<T>() {
+            private final AtomicBoolean gate = new AtomicBoolean();
+
+            @Override
+            public T receive() {
+                if (gate.getAndSet(true)) {
+                    throw new IllegalStateException("This object has already been received.");
+                }
+                return supplier.get();
+            }
+
+            @Override
+            public boolean referentIsInstanceOf(Class<?> cls) {
+                return cls.isAssignableFrom(concreteObjectType);
+            }
+
+            @Override
+            public void close() {
+                if (!gate.getAndSet(true)) {
+                    supplier.get().close();
+                }
+            }
+        };
+    }
+
+    /**
+     * Determine if the given candidate object is an instance of a {@link Send} from which an object of the given type
+     * can be received.
+     *
+     * @param type      The type of object we wish to receive.
+     * @param candidate The candidate object that might be a {@link Send} of an object of the given type.
+     * @return {@code true} if the candidate object is a {@link Send} that would deliver an object of the given type,
+     * otherwise {@code false}.
+     */
+    static boolean isSendOf(Class<?> type, Object candidate) {
+        return candidate instanceof Send && ((Send<?>) candidate).referentIsInstanceOf(type);
+    }
+
+    /**
+     * Receive the {@link Resource} instance being sent, and bind its ownership to the calling thread.
+     * The invalidation of the sent resource in the sending thread happens-before the return of this method.
+     * <p>
+     * This method can only be called once, and will throw otherwise.
+     *
+     * @return The sent resource instance.
+     * @throws IllegalStateException If this method is called more than once.
+     */
+    T receive();
+
+    /**
+     * Apply a mapping function to the object being sent. The mapping will occur when the object is received.
+     *
+     * @param type   The result type of the mapping function.
+     * @param mapper The mapping function to apply to the object being sent.
+     * @param <R>    The result type of the mapping function.
+     * @return A new {@link Send} instance that will deliver an object that is the result of the mapping.
+     */
+    default <R extends Resource<R>> Send<R> map(Class<R> type, Function<T, R> mapper) {
+        return sending(type, () -> mapper.apply(receive()));
+    }
+
+    /**
+     * Discard this {@link Send} and the object it contains.
+     * This has no effect if the send-object has already been received.
+     */
+    @Override
+    default void close() {
+        try {
+            receive().close();
+        } catch (IllegalStateException ignore) {
+            // Don't do anything if the "Send" has already been consumed.
```

Comment:
While I think what you are doing here is safe it somehow also hide an "expensive" operation as fill in the stack is expensive.